orderbook/src/orderbook.rs

```diff
@@ -78,8 +81,35 @@ impl Orderbook {
         Ok(AddOrderResult::Added(order.order_meta_data.uid))
     }
 
-    pub async fn remove_order(&self, _uid: &OrderUid) -> Result<RemoveOrderResult> {
-        todo!()
+    pub async fn cancel_order(
+        &self,
+        cancellation: &OrderCancellation,
+    ) -> Result<OrderCancellationResult> {
+        // TODO - Would like to use get_order_by_uid, but not implemented on self
+        let orders = self
+            .get_orders(&OrderFilter {
+                uid: Some(cancellation.order_uid),
+                ..Default::default()
+            })
+            .await?;
+        // Could be that order doesn't exist and is not fetched.
+        let order = match orders.first() {
+            Some(order) => order,
+            None => return Ok(OrderCancellationResult::OrderNotFound),
+        };
+
+        match cancellation.validate_signature(&self.domain_separator) {
+            Some(signer) => {
+                if signer == order.order_meta_data.owner {
+                    self.database.cancel_order(&order).await?;
+                    // Can't fail to remove here, since we know it exists already
+                    Ok(OrderCancellationResult::Cancelled)
+                } else {
+                    Ok(OrderCancellationResult::InvalidSignature)
```

Comment:
Technically the signature is valid (just for another user), but I'm fine not differentiating between the two cases (although it might help debugging if it logged the wrong address).